File: platform/android/src/main/java/com/offlinetranslator/TranslatorSession.kt
package com.offlinetranslator

import android.content.Context
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Glue for a live conversation: mic → engine (denoise + VAD + STT + NMT) → OS TTS.
 *
 *   val session = TranslatorSession(context, repo.pipelineConfig())
 *   session.onResult = { r -> showTranscript(r) }
 *   session.start(sourceLang = "auto", targetLang = "en")
 *   ...
 *   session.stop(); session.close()
 */
class TranslatorSession(
    context: Context,
    config: PipelineConfig,
    speakResults: Boolean = true,
) : AutoCloseable {

    val translator = OfflineTranslator(config)
    /** OS speech synthesis, shared with [speak]; use it to check installed voices. */
    val tts = OfflineTTSManager(context)
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private val ready = Channel<Unit>(Channel.CONFLATED)
    private var worker: Job? = null

    @Volatile var sourceLang: String = "auto"
    @Volatile var targetLang: String = "en"

    /** Speak each translation through the OS TTS. Can be toggled while running (mute). */
    @Volatile var speakResults: Boolean = speakResults

    /** Playback speed for spoken translations (1.0 = normal). */
    @Volatile var speechRate: Float = 1.0f

    var onResult: ((TranslationResult) -> Unit)? = null

    /**
     * An utterance was processed but held no speech, so [onResult] will not fire for it.
     * UIs that show a "translating" indicator need this to clear it.
     */
    var onNoSpeech: (() -> Unit)? = null
    var onError: ((String) -> Unit)? = null
    var onSpeechState: ((speaking: Boolean) -> Unit)? = null

    /**
     * Microphone loudness in 0..1, ~25×/s while capturing, for a level meter.
     * Called on the audio thread — hop to your UI dispatcher.
     */
    var onAudioLevel: ((Float) -> Unit)? = null

    private var levelFrames = 0

    private val capture = AudioCapture(context, onFrames = { pcm, n ->
        // Runs on the audio thread: cheap denoise + segmentation only.
        onAudioLevel?.let { cb ->
            if (++levelFrames >= 2) {
                levelFrames = 0
                var sum = 0.0
                for (i in 0 until n) {
                    val s = pcm[i].toDouble() / 32768.0
                    sum += s * s
                }
                val rms = if (n > 0) kotlin.math.sqrt(sum / n) else 0.0
                // Perceptual-ish curve: -50 dBFS .. 0 dBFS mapped to 0..1.
                val db = 20.0 * kotlin.math.log10(rms.coerceAtLeast(1e-6))
                cb(((db + 50.0) / 50.0).coerceIn(0.0, 1.0).toFloat())
            }
        }
        if (translator.feedAudio(pcm, n)) ready.trySend(Unit)
    }, onError = { onError?.invoke(it) })

    val isRunning: Boolean get() = capture.isRunning

    fun start(sourceLang: String = this.sourceLang, targetLang: String = this.targetLang): Boolean {
        this.sourceLang = sourceLang
        this.targetLang = targetLang
        if (capture.isRunning) return true
        translator.resetAudio()
        // Push-to-talk restarts capture many times; keep a single consumer of [ready].
        if (worker?.isActive == true) return capture.start()
        worker = scope.launch {
            for (unit in ready) {
                while (translator.pendingCount > 0) {
                    // Pause TTS so the mic does not pick up our own voice.
                    tts.stop()
                    val r = translator.processPending(this@TranslatorSession.sourceLang, this@TranslatorSession.targetLang)
                    if (!r.ok) onError?.invoke(r.error ?: "unknown error")
                    if (r.isEmpty) {
                        onNoSpeech?.invoke()
                        continue
                    }
                    onResult?.invoke(r)
                    if (r.ok && speakResults && r.translatedText.isNotBlank()) {
                        onSpeechState?.invoke(true)
                        tts.speakAndWait(r.translatedText, r.targetLang, speechRate)
                        onSpeechState?.invoke(false)
                    }
                }
            }
        }
        return capture.start()
    }

    /** Push-to-talk release: close the current utterance immediately. */
    fun endUtterance() {
        // Nothing is queued when the audio was too short or too quiet to be speech. That is not
        // an error, but a UI showing progress needs to hear about it or it waits forever.
        if (translator.flushAudio()) ready.trySend(Unit) else onNoSpeech?.invoke()
    }

    fun stop() {
        capture.stop()
        onAudioLevel?.invoke(0f)
        endUtterance()
    }

    /** Text-only path (keyboard input). */
    suspend fun translate(text: String, src: String, tgt: String): TranslationResult = withContext(Dispatchers.Default) {
        translator.translateText(text, src, tgt)
    }

    suspend fun speak(text: String, lang: String, rate: Float = speechRate) = tts.speakAndWait(text, lang, rate)

    /** Cuts off whatever the TTS is currently saying. */
    fun stopSpeaking() = tts.stop()

    override fun close() {
        stop()
        worker?.cancel()
        scope.cancel()
        tts.shutdown()
        translator.close()
    }

    companion object {
        /**
         * One-call setup: fetch the manifest, download whatever the given languages need
         * (with progress), and return a ready session. Safe to call on every launch — models
         * already installed are skipped.
         *
         *   val session = TranslatorSession.prepare(context, listOf("ko", "en")) { done, total -> ... }
         *   session.onResult = { r -> ... }
         *   session.start(sourceLang = "ko", targetLang = "en")
         */
        suspend fun prepare(
            context: Context,
            languages: List<String>,
            manifestUrl: String? = ModelRepository.DEFAULT_MANIFEST_URL,
            speakResults: Boolean = true,
            llmMode: LlmMode = LlmMode.IF_NEEDED,
            backend: TranslationBackend = TranslationBackend.AUTO,
            onProgress: ((bytesDone: Long, bytesTotal: Long) -> Unit)? = null,
        ): TranslatorSession {
            val repo = ModelRepository(context, manifestUrl = manifestUrl)
            try {
                repo.refreshManifest()
                if (!repo.hasManifest()) throw TranslatorException("no model manifest available (offline on first launch?)")
                val effectiveLlm = if (backend == TranslationBackend.LLM) LlmMode.ALWAYS else llmMode
                val needed = repo.statusForLanguages(languages, llmMode = effectiveLlm).filter { it.needsDownload }
                val total = needed.sumOf { it.totalBytes }
                if (needed.isNotEmpty()) {
                    if (repo.freeBytes() < total + 50L * 1024 * 1024) {
                        throw TranslatorException("not enough storage: need ${total / 1_000_000} MB")
                    }
                    var completedBytes = 0L
                    var failure: String? = null
                    repo.installAll(needed).collect { ev ->
                        when (ev) {
                            is InstallEvent.Downloading -> onProgress?.invoke(completedBytes + ev.bytesDone, total)
                            is InstallEvent.Installed -> {
                                completedBytes += needed.first { it.id == ev.id }.totalBytes
                                onProgress?.invoke(completedBytes, total)
                            }
                            is InstallEvent.Failed -> failure = "${ev.id}: ${ev.reason}"
                            else -> {}
                        }
                    }
                    failure?.let { throw TranslatorException("model download failed: $it") }
                }
                return TranslatorSession(context, repo.pipelineConfig(backend = backend), speakResults)
            } finally {
                repo.close()
            }
        }
    }
}

File: platform/ios/OfflineTranslator/TranslatorSession.swift
#if canImport(OfflineTranslatorObjC)
import OfflineTranslatorObjC
#endif
import AVFoundation
import Foundation

/// RMS loudness of the incoming frames, reported every other buffer (~25×/s).
/// Only ever touched from the audio render thread.
private final class LevelMeter {
    private var counter = 0

    func push(_ samples: UnsafePointer<Float>, _ count: Int) -> Float? {
        counter += 1
        guard counter >= 2, count > 0 else { return nil }
        counter = 0
        var sum: Float = 0
        for i in 0..<count { sum += samples[i] * samples[i] }
        let rms = (sum / Float(count)).squareRoot()
        // -50 dBFS ... 0 dBFS mapped to 0...1.
        let db = 20 * log10(max(rms, 1e-6))
        return min(max((db + 50) / 50, 0), 1)
    }
}

/// Live conversation glue: mic → engine (denoise + VAD + STT + NMT) → AVSpeechSynthesizer.
///
///     let session = try TranslatorSession(config: downloader.manager.pipelineConfig())
///     session.onResult = { r in print(r.sourceText, "→", r.translatedText) }
///     try session.start(sourceLang: "auto", targetLang: "en")
@MainActor
public final class TranslatorSession {

    public let pipeline: OTTranslationPipeline
    public let tts = OfflineTTSManager()
    private let capture = AudioCapture()
    private var worker: Task<Void, Never>?
    private let wakeups = AsyncStream<Void>.makeStream()

    public var sourceLang = "auto"
    public var targetLang = "en"
    /// Speak each translation through the OS TTS. Can be toggled while running (mute).
    public var speakResults = true
    /// Playback speed for spoken translations (AVSpeechUtterance rate units).
    public var speechRate: Float = AVSpeechUtteranceDefaultSpeechRate

    public var onResult: ((OTTranslationResult) -> Void)?
    public var onError: ((String) -> Void)?
    public var onSpeechState: ((Bool) -> Void)?
    /// Microphone loudness in 0...1, ~25×/s while capturing, for a level meter.
    public var onAudioLevel: ((Float) -> Void)?

    public var isRunning: Bool { capture.isRunning }

    public init(config: OTPipelineConfig) throws {
        pipeline = try OTTranslationPipeline(config: config)
        let level = LevelMeter()
        capture.onFrames = { [pipeline, wakeups, weak self] ptr, count in
            // Audio thread: denoise + segmentation only.
            if let value = level.push(ptr, count) {
                Task { @MainActor in self?.onAudioLevel?(value) }
            }
            if pipeline.feedAudio(ptr, count: UInt(count)) { wakeups.continuation.yield() }
        }
        capture.onError = { [weak self] msg in Task { @MainActor in self?.onError?(msg) } }
    }

    public func start(sourceLang: String? = nil, targetLang: String? = nil) throws {
        if let s = sourceLang { self.sourceLang = s }
        if let t = targetLang { self.targetLang = t }
        guard !capture.isRunning else { return }
        pipeline.resetAudio()

        // Push-to-talk restarts capture many times; keep a single consumer of the stream.
        if let worker, !worker.isCancelled {
            try capture.start()
            return
        }
        worker = Task.detached(priority: .userInitiated) { [weak self, pipeline, wakeups] in
            for await _ in wakeups.stream {
                while pipeline.pendingUtteranceCount > 0, !Task.isCancelled {
                    guard let self = self else { return }
                    let (src, tgt, speak, rate) = await MainActor.run { (self.sourceLang, self.targetLang, self.speakResults, self.speechRate) }
                    await MainActor.run { self.tts.stop() }
                    let r = pipeline.processPending(withSourceLang: src, targetLang: tgt)
                    await MainActor.run {
                        if !r.ok { self.onError?(r.error ?? "unknown error") }
                        if !r.isEmpty { self.onResult?(r) }
                    }
                    if r.ok, speak, !r.isEmpty, !r.translatedText.isEmpty {
                        await MainActor.run { self.onSpeechState?(true) }
                        _ = await self.tts.speakAndWait(r.translatedText, lang: r.targetLang, rate: rate)
                        await MainActor.run { self.onSpeechState?(false) }
                    }
                }
            }
        }
        try capture.start()
    }

    /// Push-to-talk release: close the current utterance immediately.
    public func endUtterance() {
        if pipeline.flushAudio() { wakeups.continuation.yield() }
    }

    public func stop() {
        capture.stop()
        onAudioLevel?(0)
        endUtterance()
    }

    /// Text-only path (keyboard input). Runs on a background thread.
    public func translate(_ text: String, from src: String, to tgt: String) async -> OTTranslationResult {
        let pipeline = self.pipeline
        return await Task.detached { pipeline.translateText(text, sourceLang: src, targetLang: tgt) }.value
    }

    /// Speaks arbitrary text (e.g. replaying a line of the transcript).
    @discardableResult
    public func speak(_ text: String, lang: String) async -> Bool {
        await tts.speakAndWait(text, lang: lang, rate: speechRate)
    }

    /// Cuts off whatever the TTS is currently saying.
    public func stopSpeaking() { tts.stop() }

    public func shutdown() {
        stop()
        worker?.cancel()
        wakeups.continuation.finish()
    }

    /// One-call setup: fetch the manifest, download what `languages` need (with progress) and
    /// return a ready session. Safe on every launch — installed models are skipped.
    ///
    ///     let session = try await TranslatorSession.prepare(languages: ["ko", "en"]) { done, total in ... }
    ///     session.onResult = { r in ... }
    ///     try session.start(sourceLang: "ko", targetLang: "en")
    public static func prepare(languages: [String],
                        manifestURL: URL? = ModelDownloader.defaultManifestURL,
                        backend: OTTranslationBackend = .auto,
                        llmMode: OTLlmMode = .ifNeeded,
                        progress: ((UInt64, UInt64) -> Void)? = nil) async throws -> TranslatorSession {
        let downloader = ModelDownloader(manifestURL: manifestURL)
        _ = await downloader.refreshManifest()
        guard downloader.hasManifest else {
            throw ModelDownloader.InstallError(message: "no model manifest available (offline on first launch?)")
        }
        let effectiveLlm: OTLlmMode = backend == .LLM ? .always : llmMode
        let needed = downloader.status(languages: languages, llmMode: effectiveLlm).filter { $0.needsDownload }
        let total = needed.reduce(UInt64(0)) { $0 + $1.totalBytes }
        if !needed.isEmpty {
            if downloader.freeBytes() < total + 50_000_000 {
                throw ModelDownloader.InstallError(message: "not enough storage: need \(total / 1_000_000) MB")
            }
            var completed: UInt64 = 0
            var failure: String?
            for await event in downloader.installAll(needed) {
                switch event {
                case .downloading(_, _, let done, _): progress?(completed + done, total)
                case .installed(let id):
                    completed += needed.first { $0.identifier == id }?.totalBytes ?? 0
                    progress?(completed, total)
                case .failed(let id, let reason): failure = "\(id): \(reason)"
                default: break
                }
            }
            if let f = failure { throw ModelDownloader.InstallError(message: "model download failed: \(f)") }
        }
        return try TranslatorSession(config: downloader.manager.pipelineConfig(with: backend))
    }
}
